### PR TITLE
Support negative values in pluck()

### LINKED ISSUE #603: Negative indices in pluck()
For indexing from the end.

### PR DESCRIPTION
**Pull request by hadley**
Fixes #603

### CODE REVIEW

File: src/pluck.c [hunk 1]
```diff
@@ -296,21 +298,29 @@ static int check_double_index_finiteness(double val, SEXP index, int i, bool str
 }
 
 static int check_double_index_length(double val, int n, int i, bool strict) {
-  if (val < 0) {
+  if (val == 0) {
+    if (strict) {
+      Rf_errorcall(R_NilValue, "Index %d is zero", i + 1);
```
Comment (lionel-) [inline src/pluck.c:303]: These error messages are missing full stops.

Comment (lionel-) [reply src/pluck.c:303]: It feels like we could do better but I'm not sure how. Maybe:

```
Error in `chuck()`:
! Problem with 3rd subscript.
Cause by error:
! Can't extract at location zero.
```

And

```
Error in `chuck()`:
! Problem with 3rd subscript.
Caused by error:
! Can't extract elements past the end.
i Location 10 doesn't exist, there are only 4 elements.

Error in `chuck()`:
! Problem with 3rd subscript.
Caused by error:
! Can't extract elements past the beginning.
i Negative location -10 doesn't exist, there are only 4 elements.
```

File: src/pluck.c [hunk 2]
```diff
@@ -296,21 +298,29 @@ static int check_double_index_finiteness(double val, SEXP index, int i, bool str
...
+  } else if (val < 0) {
     if (strict) {
+      // Negative values have already been subtracted from end
       Rf_errorcall(R_NilValue,
-                   "Index %d must be greater than 0, not %.0f",
+                   "Negative index %d must be greater than or equal to %d, not %.0f",
                    i + 1,
-                   val + 1);
+                   -n,
+                   -n + val);
```
Comment (lionel-) [inline src/pluck.c:314]:
```suggestion
                   val - n - 1);
```

**Review by lionel- — APPROVED** [review]
Nice

File: src/pluck.c (hunk 1 above)
Comment (hadley) [reply src/pluck.c:303]: See also #787

File: src/pluck.c (hunk 2 above)
Comment (hadley) [reply src/pluck.c:314]: Good catch, thanks!

**Pull request merged**